File: WithTable/app/src/main/java/com/pos_billingwala/Extra/LicenseValidator.java
package com.pos_billingwala.Extra;

import android.content.Context;
import android.provider.Settings;
import android.util.Base64;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import com.pos_billingwala.Database.POSBillingWalaDatabase;
import com.pos_billingwala.Model.LoginResponse;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.X509EncodedKeySpec;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

/**
 * Production licensing: offline validation of server-signed license payloads.
 * Public key only — private signing key never ships in the APK.
 */
public final class LicenseValidator {

    private static final String KEY_PAYLOAD = "licensePayload";
    private static final String KEY_SIGNATURE = "licenseSignature";
    private static final String KEY_LAST_SERVER_TIME = "licenseLastServerTimeMs";
    private static final long CLOCK_ROLLBACK_TOLERANCE_MS = TimeUnit.HOURS.toMillis(24);

    private LicenseValidator() {
    }

    public static void saveFromLogin(Context context, LoginResponse response) {
        if (response == null) {
            return;
        }
        Common.saveUserData(context, KEY_PAYLOAD, nullToEmpty(response.getLicensePayload()));
        Common.saveUserData(context, KEY_SIGNATURE, nullToEmpty(response.getLicenseSignature()));
        Common.saveUserData(context, "organizationId", nullToEmpty(response.getOrganizationId()));
        Common.saveUserData(context, "branchId", nullToEmpty(response.getBranchId()));
        Common.saveUserData(context, "branchLabel", nullToEmpty(response.getBranchLabel()));
        String deviceId = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.ANDROID_ID);
        Common.saveUserData(context, "deviceId", deviceId != null ? deviceId : "");
        Common.saveUserData(context, "issuedAt", nullToEmpty(response.getIssuedAt()));
        Common.saveUserData(context, "offlineGraceUntil", nullToEmpty(response.getOfflineGraceUntil()));
        Common.saveUserData(context, "trialConsumed", nullToEmpty(response.getTrialConsumed()));

        long issuedAtSec = parseLong(response.getIssuedAt(), 0L);
        if (issuedAtSec > 0L) {
            long issuedAtMs = issuedAtSec * 1000L;
            long lastKnown = parseLong(Common.getSavedUserData(context, KEY_LAST_SERVER_TIME), 0L);
            if (issuedAtMs > lastKnown) {
                Common.saveUserData(context, KEY_LAST_SERVER_TIME, String.valueOf(issuedAtMs));
            }
        }
    }

    public static boolean hasStoredPayload(Context context) {
        String payload = Common.getSavedUserData(context, KEY_PAYLOAD);
        String signature = Common.getSavedUserData(context, KEY_SIGNATURE);
        return payload != null && !payload.trim().isEmpty()
                && signature != null && !signature.trim().isEmpty();
    }

    /**
     * True when locally validated license allows billing (works offline when payload is still valid).
     */
    public static boolean isValidForBilling(Context context, POSBillingWalaDatabase database) {
        ValidationResult result = validate(context, database);
        return result.valid && !result.trialBillBlocked;
    }

    public static ValidationResult validate(Context context, POSBillingWalaDatabase database) {
        ValidationResult result = new ValidationResult();

        if (!hasStoredPayload(context)) {
            result.message = "License not activated. Connect online once to refresh.";
            return result;
        }

        SignedPayload payload = verifyAndParse(context);
        if (payload == null) {
            result.message = "Invalid license signature.";
            return result;
        }

        String deviceId = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.ANDROID_ID);
        if (deviceId == null || !deviceId.equals(payload.deviceId)) {
            result.message = "License is bound to another device.";
            return result;
        }

        String storedKey = Common.getSavedUserData(context, "LicenceKey");
        if (storedKey != null && !storedKey.isEmpty() && !storedKey.equals(payload.licenseKey)) {
            result.message = "License key mismatch.";
            return result;
        }

        long trustedNowMs = trustedNowMillis(context);
        if (detectClockRollback(context, trustedNowMs)) {
            result.message = "Device date appears incorrect. Connect online to refresh license.";
            return result;
        }

        if (!isExpiryValid(payload.expiryDate, trustedNowMs)) {
            result.message = "License expired.";
            return result;
        }

        long offlineGraceUntilMs = payload.offlineGraceUntil * 1000L;
        if (trustedNowMs > offlineGraceUntilMs) {
            result.message = "Offline license grace expired. Connect online briefly to refresh.";
            return result;
        }

        if (payload.trialConsumed == 1 || (payload.isTrial == 1 && isTrialConsumedByRules(context, database, payload))) {
            result.message = "Trial already used. Please upgrade your licence.";
            result.trialBillBlocked = true;
            return result;
        }

        if (payload.isTrial == 1 && isTrialBillCapReached(context, database, payload)) {
            result.message = blockedTrialMessage(payload);
            result.trialBillBlocked = true;
            return result;
        }

        result.valid = true;
        result.payload = payload;
        result.message = "OK";
        return result;
    }

    private static boolean isTrialConsumedByRules(Context context, POSBillingWalaDatabase database, SignedPayload payload) {
        if ("1".equals(Common.getSavedUserData(context, "trialConsumed"))) {
            return true;
        }
        return payload.trialConsumed == 1;
    }

    private static boolean isTrialBillCapReached(Context context, POSBillingWalaDatabase database, SignedPayload payload) {
        if (payload.trialMaxBills <= 0) {
            return false;
        }
        int localCount = database != null ? database.getTotalInvoiceCount() : 0;
        int serverCount = payload.trialBillCount;
        try {
            String serverRaw = Common.getSavedUserData(context, "trialBillCount");
            if (serverRaw != null && !serverRaw.trim().isEmpty()) {
                serverCount = Math.max(serverCount, Integer.parseInt(serverRaw.trim()));
            }
        } catch (NumberFormatException ignored) {
        }
        return Math.max(localCount, serverCount) >= payload.trialMaxBills;
    }

    private static String blockedTrialMessage(SignedPayload payload) {
        if (payload.trialMaxBills > 0) {
            return "Trial bill limit reached (max " + payload.trialMaxBills + "). Please upgrade your licence.";
        }
        return "Trial bill limit reached. Please upgrade your licence.";
    }

    private static SignedPayload verifyAndParse(Context context) {
        try {
            String payloadB64 = Common.getSavedUserData(context, KEY_PAYLOAD);
            String signatureB64 = Common.getSavedUserData(context, KEY_SIGNATURE);
            if (payloadB64 == null || signatureB64 == null) {
                return null;
            }

            byte[] payloadBytes = Base64.decode(payloadB64.trim(), Base64.DEFAULT);
            byte[] signatureBytes = Base64.decode(signatureB64.trim(), Base64.DEFAULT);
            String payloadJson = new String(payloadBytes, StandardCharsets.UTF_8);

            PublicKey publicKey = loadPublicKey(context);
            Signature verifier = Signature.getInstance("SHA256withRSA");
            verifier.initVerify(publicKey);
            verifier.update(payloadBytes);
            if (!verifier.verify(signatureBytes)) {
                return null;
            }

            return new Gson().fromJson(payloadJson, SignedPayload.class);
        } catch (Exception e) {
            return null;
        }
    }

    private static PublicKey loadPublicKey(Context context) throws Exception {
        InputStream in = context.getAssets().open("license_signing_public.pem");
        BufferedReader reader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
        StringBuilder pem = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            if (!line.contains("BEGIN") && !line.contains("END")) {
                pem.append(line.trim());
            }
        }
        reader.close();

        byte[] decoded = Base64.decode(pem.toString(), Base64.DEFAULT);
        X509EncodedKeySpec spec = new X509EncodedKeySpec(decoded);
        return KeyFactory.getInstance("RSA").generatePublic(spec);
    }

    private static long trustedNowMillis(Context context) {
        long deviceNow = System.currentTimeMillis();
        long lastServer = parseLong(Common.getSavedUserData(context, KEY_LAST_SERVER_TIME), 0L);
        return Math.max(deviceNow, lastServer);
    }

    private static boolean detectClockRollback(Context context, long trustedNowMs) {
        long lastServer = parseLong(Common.getSavedUserData(context, KEY_LAST_SERVER_TIME), 0L);
        if (lastServer <= 0L) {
            return false;
        }
        long deviceNow = System.currentTimeMillis();
        return deviceNow + CLOCK_ROLLBACK_TOLERANCE_MS < lastServer
                || trustedNowMs + CLOCK_ROLLBACK_TOLERANCE_MS < lastServer;
    }

    private static boolean isExpiryValid(String expiryDateYmd, long trustedNowMs) {
        if (expiryDateYmd == null || expiryDateYmd.trim().isEmpty()) {
            return false;
        }
        try {
            SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH);
            Date expiry = dateFormat.parse(expiryDateYmd.trim());
            if (expiry == null) {
                return false;
            }
            Calendar endOfExpiry = Calendar.getInstance(Locale.ENGLISH);
            endOfExpiry.setTime(expiry);
            endOfExpiry.set(Calendar.HOUR_OF_DAY, 23);
            endOfExpiry.set(Calendar.MINUTE, 59);
            endOfExpiry.set(Calendar.SECOND, 59);
            endOfExpiry.set(Calendar.MILLISECOND, 999);
            return trustedNowMs <= endOfExpiry.getTimeInMillis();
        } catch (ParseException e) {
            return false;
        }
    }

    private static long parseLong(String raw, long fallback) {
        if (raw == null || raw.trim().isEmpty()) {
            return fallback;
        }
        try {
            return Long.parseLong(raw.trim());
        } catch (NumberFormatException e) {
            return fallback;
        }
    }

    private static String nullToEmpty(String value) {
        return value != null ? value : "";
    }

    public static final class ValidationResult {
        public boolean valid;
        public boolean trialBillBlocked;
        public String message = "";
        public SignedPayload payload;
    }

    /** Mirrors server-signed JSON payload fields. */
    public static final class SignedPayload {
        public int payloadVersion;
        public String organizationId;
        public String branchId;
        public String branchLabel;
        public String licenseId;
        public String deviceId;
        public String licenseKey;
        public String licenseType;
        public int isTrial;
        public int trialMaxBills;
        public int trialBillCount;
        public int trialConsumed;
        public String expiryDate;
        public long issuedAt;
        public long offlineGraceUntil;
        public int fastBilling;
        public int takeAway;
        public int dineIn;
        public int mess;
    }
}

File: posbillingwala-flutter/lib/features/auth/domain/license_validator.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:asn1lib/asn1lib.dart';
import 'package:flutter/services.dart' show rootBundle;
import 'package:pointycastle/export.dart';
import 'package:pos_billingwala_v2/core/utils/app_platform.dart';
import 'package:pos_billingwala_v2/features/auth/domain/session_keys.dart';
import 'package:shared_preferences/shared_preferences.dart';

/* Offline RSA licence validation (Android [LicenseValidator] parity). */
class LicenseValidator {
  LicenseValidator._();

  static const payloadKey = 'licensePayload';
  static const signatureKey = 'licenseSignature';
  static const lastServerTimeKey = 'licenseLastServerTimeMs';
  static const clockToleranceMs = 24 * 60 * 60 * 1000;

  static Future<void> saveFromLogin({
    required String? licensePayload,
    required String? licenseSignature,
    required String? issuedAt,
    required String? offlineGraceUntil,
    required String? trialConsumed,
    required String? organizationId,
    required String? branchId,
    required String? branchLabel,
    required String deviceId,
  }) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(payloadKey, licensePayload?.trim() ?? '');
    await prefs.setString(signatureKey, licenseSignature?.trim() ?? '');
    await prefs.setString(SessionKeys.organizationId, organizationId ?? '');
    await prefs.setString(SessionKeys.branchId, branchId ?? '');
    await prefs.setString(SessionKeys.branchLabel, branchLabel ?? '');
    await prefs.setString('deviceId', deviceId);
    await prefs.setString('issuedAt', issuedAt ?? '');
    await prefs.setString('offlineGraceUntil', offlineGraceUntil ?? '');
    await prefs.setString('trialConsumed', trialConsumed ?? '');

    final issuedSec = int.tryParse(issuedAt?.trim() ?? '') ?? 0;
    if (issuedSec > 0) {
      final issuedMs = issuedSec * 1000;
      final last = int.tryParse(prefs.getString(lastServerTimeKey) ?? '') ?? 0;
      if (issuedMs > last) {
        await prefs.setString(lastServerTimeKey, '$issuedMs');
      }
    }

    await persistModulesFromPayload(prefs, licensePayload);
  }

  static Future<bool> hasStoredPayload() async {
    final prefs = await SharedPreferences.getInstance();
    final p = prefs.getString(payloadKey)?.trim() ?? '';
    final s = prefs.getString(signatureKey)?.trim() ?? '';
    return p.isNotEmpty && s.isNotEmpty;
  }

  static Future<LicenseValidationResult> validate({
    required String deviceId,
    required String licenceKey,
    required int localInvoiceCount,
  }) async {
    final result = LicenseValidationResult();
    final prefs = await SharedPreferences.getInstance();

    if (!await hasStoredPayload()) {
      /* No signed payload yet — fall back to expiry-date login (legacy). */
      result.valid = true;
      result.message = '';
      result.legacyFallback = true;
      return result;
    }

    final payload = await verifyAndParse(prefs);
    if (payload == null) {
      result.message = 'Licence signature invalid. Please login again.';
      return result;
    }

    if (payload.deviceId.isNotEmpty && payload.deviceId != deviceId) {
      result.message = 'Licence is bound to another device.';
      return result;
    }

    if (licenceKey.isNotEmpty &&
        payload.licenseKey.isNotEmpty &&
        licenceKey != payload.licenseKey) {
      result.message = 'Licence key mismatch.';
      return result;
    }

    final trustedNow = licenseValidatorTrustedNowMs(prefs);
    if (detectClockRollback(prefs, trustedNow)) {
      result.message = 'Device clock appears incorrect.';
      return result;
    }

    if (!isExpiryValid(payload.expiryDate, trustedNow)) {
      result.message = 'Your licence has expired. Please renew.';
      return result;
    }

    final graceMs = payload.offlineGraceUntil * 1000;
    /* Web is online-only: do not allow billing under offline grace alone. */
    if (AppPlatform.requiresNetwork) {
      /* Grace still ends the session when expired (must re-login online). */
      if (graceMs > 0 && trustedNow > graceMs) {
        result.message =
            'Session expired. Connect to the internet and login again.';
        return result;
      }
    } else if (graceMs > 0 && trustedNow > graceMs) {
      result.message = 'Offline grace period ended. Connect and login again.';
      return result;
    }

    if (payload.trialConsumed == 1 || prefs.getString('trialConsumed') == '1') {
      result.message = 'Trial already used on this licence.';
      result.trialBillBlocked = true;
      return result;
    }

    if (payload.isTrial == 1 &&
        payload.trialMaxBills > 0 &&
        localInvoiceCount >= payload.trialMaxBills) {
      result.message =
          'Trial bill limit (${payload.trialMaxBills}) reached. Please upgrade.';
      result.trialBillBlocked = true;
      return result;
    }

    result.valid = true;
    result.payload = payload;
    return result;
  }

  static Future<SignedLicensePayload?> verifyAndParse(
    SharedPreferences prefs,
  ) async {
    try {
      final payloadB64 = prefs.getString(payloadKey)?.trim();
      final signatureB64 = prefs.getString(signatureKey)?.trim();
      if (payloadB64 == null ||
          payloadB64.isEmpty ||
          signatureB64 == null ||
          signatureB64.isEmpty) {
        return null;
      }

      final payloadBytes = base64.decode(payloadB64);
      final signatureBytes = base64.decode(signatureB64);
      final publicKey = await loadPublicKey();
      final verifier = Signer('SHA-256/RSA')
        ..init(false, PublicKeyParameter<RSAPublicKey>(publicKey));
      final ok = verifier.verifySignature(
        Uint8List.fromList(payloadBytes),
        RSASignature(Uint8List.fromList(signatureBytes)),
      );
      if (!ok) return null;

      final json = jsonDecode(utf8.decode(payloadBytes));
      if (json is! Map) return null;
      return SignedLicensePayload.fromJson(Map<String, dynamic>.from(json));
    } catch (_) {
      return null;
    }
  }

  static Future<RSAPublicKey> loadPublicKey() async {
    final pem = await rootBundle.loadString(
      'assets/license_signing_public.pem',
    );
    final b64 = pem
        .replaceAll('-----BEGIN PUBLIC KEY-----', '')
        .replaceAll('-----END PUBLIC KEY-----', '')
        .replaceAll(RegExp(r'\s'), '');
    final der = base64.decode(b64);
    return parsePublicKeyFromDer(Uint8List.fromList(der));
  }

  /* Parses SubjectPublicKeyInfo (X.509) DER into [RSAPublicKey]. */
  static RSAPublicKey parsePublicKeyFromDer(Uint8List der) {
    final parser = ASN1Parser(der);
    final top = parser.nextObject() as ASN1Sequence;
    final topElements = top.elements;
    final bitString = topElements[1] as ASN1BitString;
    var keyBytes = bitString.contentBytes();
    if (keyBytes.isNotEmpty && keyBytes[0] == 0) {
      keyBytes = keyBytes.sublist(1);
    }
    final keyParser = ASN1Parser(Uint8List.fromList(keyBytes));
    final keySeq = keyParser.nextObject() as ASN1Sequence;
    final elements = keySeq.elements;
    final modulus = (elements[0] as ASN1Integer).valueAsBigInteger;
    final exponent = (elements[1] as ASN1Integer).valueAsBigInteger;
    return RSAPublicKey(modulus, exponent);
  }

  static Future<void> persistModulesFromPayload(
    SharedPreferences prefs,
    String? payloadB64,
  ) async {
    if (payloadB64 == null || payloadB64.trim().isEmpty) return;
    try {
      final bytes = base64.decode(payloadB64.trim());
      final json = jsonDecode(utf8.decode(bytes));
      if (json is! Map) return;
      final p = SignedLicensePayload.fromJson(Map<String, dynamic>.from(json));
      await prefs.setString(
        SessionKeys.fastBilling,
        p.fastBilling == 1 ? 'Yes' : 'No',
      );
      await prefs.setString(
        SessionKeys.takeAway,
        p.takeAway == 1 ? 'Yes' : 'No',
      );
      await prefs.setString(SessionKeys.dineIn, p.dineIn == 1 ? 'Yes' : 'No');
      await prefs.setString(SessionKeys.mess, p.mess == 1 ? 'Yes' : 'No');
    } catch (_) {}
  }

  static int licenseValidatorTrustedNowMs(SharedPreferences prefs) {
    final deviceNow = DateTime.now().millisecondsSinceEpoch;
    final lastServer =
        int.tryParse(prefs.getString(lastServerTimeKey) ?? '') ?? 0;
    return deviceNow > lastServer ? deviceNow : lastServer;
  }

  static bool detectClockRollback(SharedPreferences prefs, int trustedNowMs) {
    final lastServer =
        int.tryParse(prefs.getString(lastServerTimeKey) ?? '') ?? 0;
    if (lastServer <= 0) return false;
    final deviceNow = DateTime.now().millisecondsSinceEpoch;
    return deviceNow + clockToleranceMs < lastServer ||
        trustedNowMs + clockToleranceMs < lastServer;
  }

  static bool isExpiryValid(String expiryDateYmd, int trustedNowMs) {
    if (expiryDateYmd.trim().isEmpty) return false;
    try {
      final parts = expiryDateYmd.trim().split('-');
      if (parts.length < 3) return false;
      final end = DateTime(
        int.parse(parts[0]),
        int.parse(parts[1]),
        int.parse(parts[2]),
        23,
        59,
        59,
        999,
      );
      return trustedNowMs <= end.millisecondsSinceEpoch;
    } catch (_) {
      return false;
    }
  }
}

class LicenseValidationResult {
  bool valid = false;
  bool trialBillBlocked = false;
  bool legacyFallback = false;
  String message = '';
  SignedLicensePayload? payload;
}

class SignedLicensePayload {
  SignedLicensePayload({
    this.payloadVersion = 0,
    this.organizationId = '',
    this.branchId = '',
    this.branchLabel = '',
    this.licenseId = '',
    this.deviceId = '',
    this.licenseKey = '',
    this.licenseType = '',
    this.isTrial = 0,
    this.trialMaxBills = 0,
    this.trialBillCount = 0,
    this.trialConsumed = 0,
    this.expiryDate = '',
    this.issuedAt = 0,
    this.offlineGraceUntil = 0,
    this.fastBilling = 0,
    this.takeAway = 0,
    this.dineIn = 0,
    this.mess = 0,
  });

  final int payloadVersion;
  final String organizationId;
  final String branchId;
  final String branchLabel;
  final String licenseId;
  final String deviceId;
  final String licenseKey;
  final String licenseType;
  final int isTrial;
  final int trialMaxBills;
  final int trialBillCount;
  final int trialConsumed;
  final String expiryDate;
  final int issuedAt;
  final int offlineGraceUntil;
  final int fastBilling;
  final int takeAway;
  final int dineIn;
  final int mess;

  factory SignedLicensePayload.fromJson(Map<String, dynamic> json) {
    int i(dynamic v) => v is int ? v : int.tryParse(v?.toString() ?? '') ?? 0;
    String s(dynamic v) => v?.toString() ?? '';
    return SignedLicensePayload(
      payloadVersion: i(json['payloadVersion']),
      organizationId: s(json['organizationId']),
      branchId: s(json['branchId']),
      branchLabel: s(json['branchLabel']),
      licenseId: s(json['licenseId']),
      deviceId: s(json['deviceId']),
      licenseKey: s(json['licenseKey']),
      licenseType: s(json['licenseType']),
      isTrial: i(json['isTrial']),
      trialMaxBills: i(json['trialMaxBills']),
      trialBillCount: i(json['trialBillCount']),
      trialConsumed: i(json['trialConsumed']),
      expiryDate: s(json['expiryDate']),
      issuedAt: i(json['issuedAt']),
      offlineGraceUntil: i(json['offlineGraceUntil']),
      fastBilling: i(json['fastBilling']),
      takeAway: i(json['takeAway']),
      dineIn: i(json['dineIn']),
      mess: i(json['mess']),
    );
  }
}
